### PR TITLE
WIP - fix: trim so name when is longer than 63 chars

### PR DESCRIPTION
**Pull request by JorTurFer**
Signed-off-by: Jorge Turrado <jorge_turrado@hotmail.es>

If the SO Name is longer than 63 characters, KEDA fails because that length is valid for the SO Name but not for other resources where it's used. This PR trims the SO Name when is longer than 63 characters on every place where is needed

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes https://github.com/kedacore/keda/issues/2915

### CODE REVIEW

**Comment by tomkerkhove** [comment]
How do you resolve it when `verylongnaaa[..]aame-1` & `verylongnaaa[..]aame-2` are 65 chars long?

**Comment by JorTurFer** [comment]
In than case there is a conflict and the hpa will not be created. We can remove the extra characters from the beginning or from the middle, but there are cases for every situation.
Other option to simplify this is just throw an error if the scaled object is longer than IDK, 55 characters (remember that the add things to the so name in some cases so we extend the length)

**Comment by JorTurFer** [comment]
Or we could also remove the odd characters or something like that 🤔

**Comment by tomkerkhove** [comment]
Or we assign a guid internally and use that but I guess that defeats the discoverability for our end-users - unless it's surfaced in our k get scaledobject

**Comment by JorTurFer** [comment]
> Or we assign a guid internally and use that but I guess that defeats the discoverability for our end-users - unless it's surfaced in our k get scaledobject

You are right, it's another option, but I agree that the user experience will be horrible in that case. The main problem is that it's not only an internal thing (like label selectors), the HPA name is generated based on SO Name and it has a length constraint AFAIK

**Comment by zroubalik** [comment]
As mentioned, this could introdue conflict when there will be a couple of SO with a similar long name. Also the label set on ScaledObject is being used to identify correct SO that is needed by HPA controller to drive the scaling.

We should design some robust solution (probably introducing some uids), but I think that we don't have to block 2.7 on this and postpone it to 2.8

**Comment by tomkerkhove** [comment]
Agreed - I think unique IDs are the only solid approach for this which we can surface in our end-user experience.

**Comment by JorTurFer** [comment]
I agree too. Let's talk about this with calm for v2.8

File: controllers/keda/hpa.go [hunk 1]
```diff
@@ -73,18 +71,12 @@ func (r *ScaledObjectReconciler) newHPAForScaledObject(ctx context.Context, logg
...
-		labelName = strings.TrimRightFunc(labelName, func(r rune) bool {
-			return !unicode.IsLetter(r) && !unicode.IsNumber(r)
-		})
-	}
+	// label can have max 63 chars
 	labels := map[string]string{
 		"app.kubernetes.io/name":       labelName,
 		"app.kubernetes.io/version":    version.Version,
-		"app.kubernetes.io/part-of":    scaledObject.Name,
+		"app.kubernetes.io/part-of":    kedacontrollerutil.Trim(scaledObject.Name, 63),
```
Comment (tomkerkhove) [inline controllers/keda/hpa.go:79]: As discussed, we need a more robust solution

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
Just for sake of process

**Comment by rwkarg** [comment]
Another approach I've seen is if the string is over the limit (ex. >63), cut it down to a smaller number (63-7=56) and add a hash of the full original value (`-` plus a 6 character hash). 
So `long-value-...-the-rest-of-the-thing` becomes `long-value-...-the-re-39f5a2`.

Doesn't guarantee there won't be a collision, but it's extremely unlikely.

**Comment by zroubalik** [comment]
Agree, I have been thinking about something like this.

We need to evaluate all the options and select the best  one. We still have a time for 2.8

**Pull request closed without merging**